### PR TITLE
Add consider-if-expression extension

### LINKED ISSUE #4366: New check: Use ternary for single assignment
A variable gets a value assigned conditionally:

```python
if condition():
    x = 4:
else:
    x = 5
```

I don't like blocks like this because they make the situation look
more complex than it really is. A single variable is getting a value
assigned. What that value is is conditional, but the assignment
itself is not. There's one variable and one assignment, so there
should be just one statement with a ternary expression:

```python
x = 4 if condition() else 5
```

This is a rather opinionated check, so it should be added as an
extension rather than enabled by default.

### PR DESCRIPTION
**Pull request by nickdrozd**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #4366

### CODE REVIEW

**Comment by nickdrozd** [comment]
Something is off with the functional test specification. Am I missing something obvious?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Hey, thank you for implementing this. Could you revert all the ternary changes that do not fit on one line using black ? I think bigger ternary make the code harder to understand.

**Comment by cdce8p** [comment]
> Hey, thank you for implementing this. Could you revert all the ternary changes that do not fit on one line using black ? I think bigger ternary make the code harder to understand.

I agree with that. Ternary ifs should only really be used for short conditions, something like
```py
var = 2 if some_var else 3
```

**Comment by Pierre-Sassoulas** [comment]
To the point that I'm wondering if the checker should not be changed to :
- if it's a ternary and it's longer than 88 characters, tell the user to remove the ternary
- It's it's not a ternary and it's shorter than 88 characters, tell the user to consider a ternary

What do you think @nickdrozd  ?

**Comment by cdce8p** [comment]
> To the point that I'm wondering if the checker should not be changed to :
> 
> * if it's a ternary and it's longer than 88 characters, tell the user to remove the ternary
> * It's it's not a ternary and it's shorter than 88 characters, tell the user to consider a ternary

I'm really not sure we should enforce any of it. IMO it doesn't make the code inherently better to use either.
I.e. even for the existing pylint code I don't think we should do most of these changes.

**Comment by nickdrozd** [comment]
Okay, I took out the changes that don't result in one-liners.

Personally, I would change all of them. It's not about lines of text, it's about complexity -- one single assign statement is less complex than two. It's a shame that Python makes if-expressions so clunky to use. In Lisp and Rust if-expressions are used all over the place with no problems.

As for the check itself, I think it's okay if it suggests some more extreme changes, like if-expressions that would get broken across multiple lines. The check is an extension, disabled by default. So anyone who enables the extension can be safely assumed to be a relatively sophisticated Pylint user, and therefore also someone who can deal with any false positives or iffy cases.

File: pylint/extensions/consider_if_expression.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from astroid import nodes
+
+from pylint.checkers import BaseChecker
+from pylint.interfaces import IAstroidChecker
+
+
+class ConsiderIfExpressionChecker(BaseChecker):
+
+    __implements__ = (IAstroidChecker,)
+    name = "consider_if_expression"
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/consider_if_expression.py:12]:
```suggestion
    name = "consider_ternary_expression"
```

File: pylint/extensions/consider_if_expression.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    msgs = {
+        "W0160": (
+            "Consider rewriting as an if-expression",
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/consider_if_expression.py:15]:
```suggestion
            "Consider rewriting as a ternary if-expression",
```

File: pylint/extensions/consider_if_expression.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            "consider-if-expression",
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/consider_if_expression.py:16]:
```suggestion
            "consider-ternary-expression",
```

File: pylint/extensions/consider_if_expression.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            "Multiple assign statements spread across if/else blocks can be "
+            "rewritten with a single assingment and an if-expression",
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/consider_if_expression.py:18]:
```suggestion
            "rewritten with a single assignment as a ternary conditional",
```

File: pylint/extensions/consider_if_expression.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        )
+    }
+
+    def visit_if(self, node):
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/consider_if_expression.py:22]: Could you add the typing, please ? :)

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for cleaning up the MR, I left some comment, mostly I think the expected name is "ternary operator" or "ternary conditional".

**Comment by Pierre-Sassoulas** [comment]
> As for the check itself, I think it's okay if it suggests some more extreme changes,

Yes I agree, I was trying to transform it into something complicated that I would personally use, but if someone want only ternary and have to activate this plugin, then it can be very opinionated without problems.

**Review by Pierre-Sassoulas — APPROVED** [review]
Great work, thank you !

**Comment by zrothberg** [comment]
https://nickdrozd.github.io/2021/09/02/new-pylint-checks.html

This dude just used your repo to write a terrible article.

**Pull request merged**